File: luaui/geom/Point.lua
local Point = {}
Point.__index = Point

function Point.new(x, y)
    local self = setmetatable({}, Point)
    self.x = x or 0
    self.y = y or 0
    return self
end

function Point:add(p) self.x = self.x + p.x; self.y = self.y + p.y end
function Point:subtract(p) self.x = self.x - p.x; self.y = self.y - p.y end
function Point:length() return math.sqrt(self.x * self.x + self.y * self.y) end
function Point:copy() return Point.new(self.x, self.y) end

return Point

File: flutter_ui/lib/geom/point.dart
import 'dart:math' as math;

class Point {
  double x;
  double y;

  Point([this.x = 0, this.y = 0]);

  void set([double x = 0, double y = 0]) {
    this.x = x;
    this.y = y;
  }

  double length() => math.sqrt(x * x + y * y);

  void multiply(double factor) {
    x *= factor;
    y *= factor;
  }

  Point product(double factor) => Point(x * factor, y * factor);

  void normalize(double targetLength) {
    if (x == 0 && y == 0) return;
    double norm = targetLength / length();
    multiply(norm);
  }

  Point normalized(double targetLength) {
    if (x == 0 && y == 0) return Point();
    double norm = targetLength / length();
    return product(norm);
  }

  Point orth() => normalized(1);

  Point copy() => Point(x, y);
}
